nutils/function.py

```diff
@@ -147,9 +140,16 @@ def __getitem__(self, space: str) -> LowerArg:
 
     @property
     def spaces(self):
-        'The spaces of the :attr:`args`'
+        '''The spaces of the :attr:`args` without repetition
 
-        return tuple(arg.space for arg in self.args)
+        The spaces are ordered by :attr:`args`. Spaces that have another entry
+        in :attr:`args` to the right are omitted.
+        '''
+
+        spaces = []
+        spaces.extend(arg.space for arg in reversed(self.args) if arg.space not in spaces)
+        spaces.reverse()
+        return tuple(spaces)
```

Comment:
In the situations where the order is used, the next step seems to always be take take the argument from args via args[space], so maybe we should offer something like .exposed_args instead for these situations. Then for spaces we can simply return `{arg.space for arg in self.args}`.

Reply:
I've add `LowerArgs.exposed` and changed `LowerArgs.spaces` to a `frozenset`, since we're now only using this attribute to test existence of certain spaces (in `_Opposite`).